packages/report-server/src/__tests__/reportBuilder/transform/insert.test.ts

```diff
@@ -0,0 +1,215 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { SINGLE_ANALYTIC, MULTIPLE_ANALYTICS, AGGREGATEABLE_ANALYTICS } from './transform.fixtures';
+import { buildTransform } from '../../../reportBuilder/transform';
+
+describe('insert', () => {
+  // SAME AS SELECT FUNCTIONALITY
+  it('can insert row with basic values', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'number'": '1',
+        "'string'": "'Hi'",
+        "'boolean'": 'false',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { number: 1, string: 'Hi', boolean: false }]);
+  });
+
+  it('can insert row with values from previous row', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { dataElementValue: 4 }]);
+  });
+
+  it('can select a value from the row as a field name', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        '$row.dataElement': '$row.value',
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { BCD1: 4 }]);
+  });
+
+  it('can execute functions', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'period'": "periodToDisplayString($row.period, 'DAY')",
+      },
+    ]);
+    expect(transform(SINGLE_ANALYTIC)).toEqual([...SINGLE_ANALYTIC, { period: '1st Jan 2020' }]);
+  });
+
+  // SPECIFIC TO INSERT
+  it('can insert a single row at start', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        position: 'before',
+        where: 'eq($index, 0)',
```

Comment:
How would you feel about us using a 1-based index rather than a 0-based index? I know it goes against our creed, but I'm partial towards it since we plan on displaying the row# in the data table preview in the upcoming Admin Panel Viz Builder: 
![image](https://user-images.githubusercontent.com/59544282/126265525-6aa5349c-2125-42e8-9efa-1d587214a1fa.png)

Probably not a decision we need to make right away tho 👍

Reply:
Personally I think it would make more sense to make the front end zero indexed. 🤔

If it's visualising data it makes sense to keep it the way we use it in data. Plus it means there's less places to have an off by one error.

Reply:
I would definitely agree with you @EMcQ-BES if our audience was going to stay internal devs forever...but the idea with the viz builder is that we could open it up to non devs. Probably after the solutions team the next lot of users would be people who are savvy in Excel, and they're likely to be much more familiar with 1-indexing than 0-indexing. So from the UX perspective, I'm keen to keep the front end starting at 1

The question then is where we want to translate that, and because we're exposing the raw transform step code to the user in the same interface as the preview table, I'd probably also be in favour of making it 1-indexed there too.

If the UI was fully taking care of holding the user's hand when writing "where" statements etc., and giving them options through a drop down etc., maybe we could keep the transform step 0-indexed and make the UI do the translation, but we're a long way off having an interface that is that tight.

Reply:
Actually I might be on your side... attempting to use `$table[0]` causes an error because it's apparently one-based... I'm not sure which point of the process makes that happen but currently it means `$table[$index]` is not equal to `$row` and that definitely feels incorrect.

Reply:
I've adjusted `$index` to be one-based

Reply:
Few things to note:
> Probably after the solutions team the next lot of users would be people who are savvy in Excel, and they're likely to be much more familiar with 1-indexing than 0-indexing.

This is how I see it too! And tbh even as a dev, I still enjoy 'tables' as being 1-indexed. The fact that a table is an array of rows seems a bit like an implementation detail 👍 On that note I'm somewhat keen for us to steer away from having arrays as a construct in the report-server expressions; ie. use `sum(a, b, c)` instead of `sum([a,b,c])`. This is for a couple reasons:
- Arrays are quite a programmer-y concept and probably not so familiar to our hopefully non-dev audience
- mathjs matrices and arrays aren't quite the same, and I've been caught out by bugs where I thought I was dealing with an array when in fact it was a mathjs matrix. At the same time, I don't wanna lean too heavily into using mathjs matrices wherever we might want to use arrays as they seem inappropriate much of the time, so best to avoid all together?

Regarding that point, I absolutely love your `length()` function @EMcQ-BES! It reminds me of python's `len()` function which worked for arrays, strings, dictionaries (objects), you name it. Python was the first language I learned and was very beginner friendly! I think if we extended this function (and others too) to be polymorphic over time, it would save our users from needing to think about data structures 🎉